From: Daniel van Vugt <daniel.van.vugt@canonical.com>
Date: Mon, 13 Jul 2026 18:07:22 +0800
Subject: Multi-monitor magnifier fixes

Support for multiple monitors in the magnifier had bitrotted, if it ever
worked at all in Wayland sessions. This fixes clipping and sizing issues.

Bug: https://gitlab.gnome.org/GNOME/gnome-shell/-/work_items/8101
Bug-Ubuntu: https://bugs.launchpad.net/bugs/2108011
---
 js/ui/background.js |  2 +-
 js/ui/magnifier.js  | 15 ++++++---------
 2 files changed, 7 insertions(+), 10 deletions(-)

diff --git a/js/ui/background.js b/js/ui/background.js
index 25b4f18..c0ddc53 100644
--- a/js/ui/background.js
+++ b/js/ui/background.js
@@ -108,7 +108,7 @@ import * as LoginManager from '../misc/loginManager.js';
 import * as Main from './main.js';
 import * as Params from '../misc/params.js';
 
-const DEFAULT_BACKGROUND_COLOR = new Cogl.Color({red: 40, green: 40, blue: 40, alpha: 255});
+export const DEFAULT_BACKGROUND_COLOR = new Cogl.Color({red: 40, green: 40, blue: 40, alpha: 255});
 
 const BACKGROUND_SCHEMA = 'org.gnome.desktop.background';
 const PRIMARY_COLOR_KEY = 'primary-color';
diff --git a/js/ui/magnifier.js b/js/ui/magnifier.js
index 9da1f77..3707589 100644
--- a/js/ui/magnifier.js
+++ b/js/ui/magnifier.js
@@ -797,7 +797,6 @@ class ZoomRegion {
         this._contrast = {r: NO_CHANGE, g: NO_CHANGE, b: NO_CHANGE};
 
         this._magView = null;
-        this._background = null;
         this._uiGroupClone = null;
         this._mouseSourceActor = mouseSourceActor;
         this._mouseActor  = null;
@@ -1494,23 +1493,23 @@ class ZoomRegion {
 
     _createActors() {
         // Add a group to clip the contents of the magnified view.
-        const mainGroup = new Clutter.Actor({clip_to_allocation: true});
+        const mainGroup = new Clutter.Actor({
+            clip_to_allocation: true,
+            x_expand: true,
+            y_expand: true,
+        });
 
         // The root actor for the zoom region
         this._magView = new St.Bin({
             style_class: 'magnifier-zoom-region',
             child: mainGroup,
+            background_color: Background.DEFAULT_BACKGROUND_COLOR,
         });
         global.stage.add_child(this._magView);
 
         // hide the magnified region from CLUTTER_PICK_ALL
         Shell.util_set_hidden_from_pick(this._magView, true);
 
-        // Add a background for when the magnified uiGroup is scrolled
-        // out of view (don't want to see desktop showing through).
-        this._background = new Background.SystemBackground();
-        mainGroup.add_child(this._background);
-
         // Clone the group that contains all of UI on the screen.  This is the
         // chrome, the windows, etc.
         this._uiGroupClone = new Clutter.Clone({
@@ -1550,7 +1549,6 @@ class ZoomRegion {
         this._magShaderEffects = null;
         this._magView.destroy();
         this._magView = null;
-        this._background = null;
         this._uiGroupClone = null;
         this._mouseActor = null;
         this._crossHairsActor = null;
@@ -1837,7 +1835,6 @@ class ZoomRegion {
     }
 
     _monitorsChanged() {
-        this._background.set_size(global.screen_width, global.screen_height);
         this._updateScreenPosition();
     }
 }
